### ISSUE TITLE
QCSchema with PBC?

### ISSUE THREAD

**Issue by awvwgk** [MolSSI/QCSchema#66]
I looked at the specifications of the QCSchema at
https://molssi-qc-schema.readthedocs.io/en/latest/auto_topology.html
and noticed that I cannot pass lattice parameters or the systems periodicity with the molecules topology.

Are you planning to add periodic boundary conditions to QCSchema?

**Comment by dgasmith** [comment]
This has always seemed a bit out of scope to me. There was some strong push for it originally, but it seems that Materials Project, AIDIIA, Nomad, ASE, etc cover this space fairly well.

I wouldn't rule it out entirely, but I think there would need to be some strong use cases for it.

**Comment by awvwgk** [comment]
Yes, there are competing standards out there and one has to support more than one in the end.
So it is important to know if I can (potentially) use QCSchema in the programs I am developing, which necessarily requires having PBC.

**Comment by dgasmith** [comment]
@loriab @wadejong Any thoughts here?

**Comment by wadejong** [comment]
Relevant for molecular dynamics and materials science. But, it’s becoming also relevant in the self assembly research of mixed organic-inorganic molecular assemblies. I would say it is and should be in scope.

**Comment by dgasmith** [comment]
@mkhorton does MP have schema that we can borrow for this?

**Comment by ghutchis** [comment]
I think it would be good to have at least PBC / unit vectors.

**Comment by cryos** [comment]
So do I, we added them in Chemical JSON quite some time ago. We have never mixed orbitals and periodic boundary conditions together.

**Comment by dgasmith** [comment]
@cryos Could you post that schema here or go ahead and make a PR?

**Comment by awvwgk** [comment]
I found chemical JSON is using a representation with cell parameters.

```json
"unit cell": {"a": 9.1, "b": 9.1, "c": 30.5, "alpha": 90, "beta": 90, "gamma": 120}
```

This format allows to easily specify 1D and 2D periodic systems and possibly spacegroup symmetry.

A band structure might be a bit more complicated and could be out-of-scope, but the stress tensor would be nice to have in the output.

**Comment by ghutchis** [comment]
Had to go check the code, the syntax should be:

`"cellVectors": { 10.5, 0.0, 0.0, 0.0, 10.5, 0.0, 0.0, 0.0, 10.5 }`

In other words, the keyword `cellVectors` and then 9 distances in Angstrom.

**Comment by mkhorton** [comment]
In the Materials Project JSON we use:

```
'lattice': {'matrix': [[2.32547694, 0.0, -0.82218026],
   [-1.16273847, 2.01392211, -0.82218026],
   [0.0, 0.0, 2.46654077]],
  'a': 2.4665407716892562,
  'b': 2.4665407749920663,
  'c': 2.46654077,
  'alpha': 109.47122067561966,
  'beta': 109.47122070274486,
  'gamma': 109.47122053911131,
  'volume': 11.55162296798055},
```

The canonical information is the `matrix` since this unambiguously includes an orientation also, the additional information (a, b, c, alpha, beta, gamma, volume) are solely for human readability.

**Comment by mkhorton** [comment]
One additional factor to consider with PBC is whether you want to specific PBC along some directions only. Mixed boundary conditions present challenges, and we do not support them, but some tools (LAMMPS and ASE) do use them.

**Comment by mkhorton** [comment]
Finally, I think I mentioned this previously but it's an important point, if your schema can include information on chemical bonds (or any other kind of connectivity/"edge" information between atomic sites), then you have to think about how this information is stored with periodic boundaries also. For example, if you have a bond from your original unit cell bonded to an atom in the next periodic image cell over.

**Comment by awvwgk** [comment]
I'm bumping this issue, since I'm still interested in having PBCs in QCSchema.

At the ASE developer workshop we shortly discussed PBCs and topology and they seems to be no problem with having both. Here is the merge request from their gitlab for reference:
https://gitlab.com/ase/ase/merge_requests/1321

**Comment by awvwgk** [comment]
There was only an issue for connecting an atom with its own image, of course. But this is a pretty special case for very small unit cells.

**Comment by mkhorton** [comment]
> But this is a pretty special case for very small unit cells.

Yes, agreed -- and a very important case if you deal with a lot of small unit cells. In pymatgen we've ended up storing the destination lattice vector (e.g. (1, 0, 0) for the next periodic image in the +x direction) along with the atom pair defining the bond in our serialization format. Regardless of implementation though, there's definitely a need for it.

**Comment by dgasmith** [comment]
Can anyone make a PR here to put these discussions into practice?

**Comment by awvwgk** [comment]
Sorry for reviving this old thread. How would we implement this in QCSchema, I'm a bit unsure about the actual workflow here.

As far as I understood this repo is output from QCElemental rather than a standalone project which is implemented in QCElemental, right? Implementing this in QCElemental would therefore be the first step?

**Comment by ghutchis** [comment]
@awvwgk - many of us in the community pushed very strongly that this was the opposite. Discussion should happen here. If an implementation happens in QCElemental, it should be later.

(There are many of us following this repo and most of the discussion threads have been here.)

**Comment by ghutchis** [comment]
I think I'd favor something similar to @mkhorton and the Materials Project schema. The lattice matrix is the most important and unambiguous if available, with a, b, c, alpha, beta, gamma being useful for humans or a quick glance.

**Comment by berquist** [comment]
> As far as I understood this repo is output from QCElemental rather than a standalone project which is implemented in QCElemental, right? Implementing this in QCElemental would therefore be the first step?

We are waiting on https://github.com/MolSSI/QCSchema/pull/77 and related things marked there, if the final decision is to generate from QCElemental.

**Comment by awvwgk** [comment]
I am implementing a proposal for periodic boundary conditions in QCSchema with https://github.com/grimme-lab/mctc-lib/pull/93. The format will use the extras table and add the lattice vectors in `extras.periodic.lattice` as a flattened array in Bohr. I would be happy to get feedback on this proposal.

This will be used and available in packages like [`xtb`](https://github.com/grimme-lab/xtb), [`tblite`](https://tblite.readthedocs.io), [`dftd4`](https://dftd4.readthedocs.io), and [`s-dftd3`](https://dftd3.readthedocs.io).

Here is an example of using QCSchema output for periodic systems with the proposed format:

<details><summary>TiO2 in QCSchema format</summary>

````json
{
  "schema_version": 2,
  "schema_name": "qcschema_molecule",
  "provenance": {
    "creator": "mctc-lib",
    "version": "0.4.2",
    "routine": "mctc_io_write_qcschema::write_qcschema"
  },
  "comment": "TiO2 rutile",
  "symbols": ["Ti", "Ti", "O", "O", "O", "O"],
  "atomic_numbers": [22, 22, 8, 8, 8, 8],
  "geometry": [
     0.0000000000000000E+00, 0.0000000000000000E+00, 0.0000000000000000E+00,
     5.2818191416515159E+00, 8.2022538117381334E+00, 8.2022538117381334E+00,
     6.1333938828927657E-16, 5.0082961774473045E+00, 5.0082961774473045E+00,
     1.3956333869785798E-15, 1.1396211446028962E+01, 1.1396211446028962E+01,
     5.2818191416515150E+00, 3.1939576342908298E+00, 1.3210549989185438E+01,
     5.2818191416515150E+00, 1.3210549989185438E+01, 3.1939576342908289E+00
  ],
  "molecular_charge": 0,
  "extras": {
    "periodic": {
      "lattice": [
         5.5900366437622173E+00, 0.0000000000000000E+00, 0.0000000000000000E+00,
         5.3155130499965102E-16, 8.6808915904526547E+00, 0.0000000000000000E+00,
         5.3155130499965102E-16, 5.3155130499965102E-16, 8.6808915904526547E+00
      ]
    }
  }
}
````
</details>

For comparison I included a few other formats with support for periodic boundary conditions here:

<details><summary>TiO2 in Chemical JSON format</summary>

````json
{
  "chemicalJson": 1,
  "name": "TiO2 rutile",
  "atoms": {
    "elements": {"number": [22, 22, 8, 8, 8, 8]},
    "coords": {
      "3d": [
         0.0000000000000000E+00, 0.0000000000000000E+00, 0.0000000000000000E+00,
         2.7950183218811095E+00, 4.3404457952263273E+00, 4.3404457952263273E+00,
         3.2456522683278691E-16, 2.6502762025651956E+00, 2.6502762025651956E+00,
         7.3853738316651503E-16, 6.0306153878874591E+00, 6.0306153878874591E+00,
         2.7950183218811091E+00, 1.6901695926611322E+00, 6.9907219977915229E+00,
         2.7950183218811091E+00, 6.9907219977915229E+00, 1.6901695926611318E+00
      ]
    }
  },
  "properties": {"totalCharge": 0},
  "unitCell": {
    "a":  2.9581200000000001E+00,
    "b":  4.5937299999999990E+00,
    "c":  4.5937299999999990E+00,
    "alpha":  9.0000000000000000E+01,
    "beta":  9.0000000000000000E+01,
    "gamma":  9.0000000000000000E+01
  }
}
````
</details>

<details><summary>TiO2 in Vasp format</summary>

````
TiO2 rutile
    1.00000000000000
    5.59003664376222    0.00000000000000    0.00000000000000
    0.00000000000000    8.68089159045265    0.00000000000000
    0.00000000000000    0.00000000000000    8.68089159045265
 Ti   O   
 2 4
Cartesian
    0.00000000000000    0.00000000000000    0.00000000000000
    2.79501832188111    4.34044579522633    4.34044579522633
    0.00000000000000    2.65027620256520    2.65027620256520
    0.00000000000000    6.03061538788746    6.03061538788746
    2.79501832188111    1.69016959266113    6.99072199779152
    2.79501832188111    6.99072199779152    1.69016959266113
````
</details>

<details><summary>TiO2 in DFTB+ gen format</summary>

````
6 S
 Ti O
    1    1    0.00000000000000E+00    0.00000000000000E+00    0.00000000000000E+00
    2    1    2.79501832188111E+00    4.34044579522633E+00    4.34044579522633E+00
    3    2    3.24565226832787E-16    2.65027620256520E+00    2.65027620256520E+00
    4    2    7.38537383166515E-16    6.03061538788746E+00    6.03061538788746E+00
    5    2    2.79501832188111E+00    1.69016959266113E+00    6.99072199779152E+00
    6    2    2.79501832188111E+00    6.99072199779152E+00    1.69016959266113E+00
    0.00000000000000    0.00000000000000    0.00000000000000
    2.95812000000000    0.00000000000000    0.00000000000000
    0.00000000000000    4.59373000000000    0.00000000000000
    0.00000000000000    0.00000000000000    4.59373000000000
````
</details>

<details><summary>TiO2 in Turbomole coord format</summary>

````
$coord
        0.00000000000000        0.00000000000000        0.00000000000000      Ti
        5.28181914165152        8.20225381173813        8.20225381173813      Ti
        0.00000000000000        5.00829617744730        5.00829617744730      O
        0.00000000000000       11.39621144602896       11.39621144602896      O
        5.28181914165152        3.19395763429083       13.21054998918544      O
        5.28181914165152       13.21054998918544        3.19395763429083      O
$periodic 3
$lattice
    5.59003664376222    0.00000000000000    0.00000000000000
    0.00000000000000    8.68089159045265    0.00000000000000
    0.00000000000000    0.00000000000000    8.68089159045265
$end
````
</details>

<details><summary>TiO2 in FHI-Aims format</summary>

````
atom         0.00000000000000        0.00000000000000        0.00000000000000 Ti
atom         2.79501832188111        4.34044579522633        4.34044579522633 Ti
atom         0.00000000000000        2.65027620256520        2.65027620256520 O
atom         0.00000000000000        6.03061538788746        6.03061538788746 O
atom         2.79501832188111        1.69016959266113        6.99072199779152 O
atom         2.79501832188111        6.99072199779152        1.69016959266113 O
lattice_vector         2.95812000000000        0.00000000000000        0.00000000000000
lattice_vector         0.00000000000000        4.59373000000000        0.00000000000000
lattice_vector         0.00000000000000        0.00000000000000        4.59373000000000
````
</details>
